### ISSUE TITLE
Experimental VJP cache can return another input's gradient, and breaks apply for non-JAX inputs

### ISSUE THREAD

**Issue by singhharsh1708** [pasteurlabs/tesseract-core#666]
The experimental VJP cache (`set_jax_vjp_cache_size`) has two independent problems. The first returns a wrong gradient with no error; the second makes `apply` fail on a Tesseract that works fine with the cache off.

## 1. A cache-key collision serves another input's backward pass

`_hash_tree` collapses the key with `hash()` (`jax_recipes.py:58`), and `LRUCache.get` is a bare dict lookup with no stored key to compare against (`tree_transforms.py:206-212`). The int *is* the whole key, so any hash collision is a silent wrong gradient.

Non-array leaves are appended raw, so the key inherits CPython's hash identities. CPython reserves `-1` as an error sentinel, so `hash(-1) == hash(-2) == -2`.

This reproduces on the shipped `examples/vectoradd_jax`, whose `norm_ord: int = 2` is passed to `jnp.linalg.norm(ord=...)`, where `-1` and `-2` are both valid orders:

```
cache disabled (ground truth)
  vjp(norm_ord=-1): [ 0.2539683  -0.1020408  -0.06172839]
  vjp(norm_ord=-2): [ 0.12143232 -0.05404405 -0.02542813]

cache enabled, apply(norm_ord=-1) then vjp(norm_ord=-2)
  returned:         [ 0.2539683  -0.1020408  -0.06172839]   <- the norm_ord=-1 backward
```

There is a second direction too: `1 == 1.0 == True` compare equal *and* hash equal, so the value alone cannot discriminate a leaf whose field is typed as a union.

## 2. Enabling the cache breaks any Tesseract with a non-JAX input

`jax_apply` hands the entire input dict to `jax.vjp` (`jax_recipes.py:99-101`), which traces every leaf it receives. A `str` or `None` input therefore aborts the call:

```
cache OFF: apply OK -> [2. 4. 6.]
cache ON : apply RAISED InvalidInputException: Argument 'fast' of type <class 'str'> is not a valid JAX type.
```

The uncached path is fine, because `filter_func` threads non-differentiable leaves in as closure constants.

## Why the existing tests do not catch either

`TestCacheCorrectness::test_cache_on_matches_cache_off` is the right test in spirit, but `_build_api`'s schema carries only float arrays and float scalars, and `_make_inputs` varies inputs by adding an offset to the *array* values. Array leaves are keyed on their bytes and discriminate correctly, so the collision path is never exercised; the one scalar assertion uses 42 vs 43, which do not collide. And no leaf in that schema is a type `jax.vjp` refuses to trace.

## Fix

For (1), return the key tuple instead of collapsing it, so the dict compares with `__eq__` rather than trusting the hash, and tag non-array leaves with their type. Note that swapping in a stronger digest would not fix this: it would still be a hash used as a key with no equality check, just with a smaller collision probability.

For (2), partition the inputs with `eqx.partition` before `jax.vjp` and thread the static leaves back in, mirroring what `jax_abstract_eval` already does at `jax_recipes.py:187-193`. The filter wants to be `eqx.is_inexact_array_like` rather than `eqx.is_array`, otherwise a scalar `Differentiable[Float32]` field becomes static and silently loses its gradient.

I have both fixes with regression tests and will open a PR shortly.
